fedimint-server/src/consensus/engine.rs

```diff
@@ -472,6 +484,81 @@ impl ConsensusEngine {
             .expect("This is the only place where we write to this key");
     }
 
+    /// Creates the directory within the data directory for storing the database
+    /// checkpoints.
+    fn create_checkpoint_directory(&self) -> anyhow::Result<()> {
+        let checkpoint_dir = self.data_dir.join("db_checkpoints");
+
+        if !checkpoint_dir.exists() {
+            fs::create_dir_all(&checkpoint_dir)?;
+        }
+
+        Ok(())
+    }
+
+    /// Creates a backup of the database in the checkpoint directory. The
+    /// frequency of the checkpoints is controlled by
+    /// `FM_DB_CHECKPOINT_FREQ_ENV`. These checkpoints can be used to
+    /// restore the database in case the federation falls out of consensus.
+    fn checkpoint_database(&self, session_index: u64) {
+        // By default, checkpoint the database every 10 sessions
+        let checkpoint_frequency: u64 = env::var(FM_DB_CHECKPOINT_FREQ_ENV)
+            .unwrap_or(10.to_string())
+            .parse()
+            .expect("FM_DB_CHECKPOINT_FREQ var is invalid");
+
+        if session_index % checkpoint_frequency == 0 {
+            let checkpoint_dir = self.data_dir.join("db_checkpoints");
+            let session_checkpoint_dir = checkpoint_dir.join(format!("{session_index}"));
+
+            {
+                let _timing /* logs on drop */ = timing::TimeReporter::new("database-checkpoint").info();
+                match self.db.checkpoint(&session_checkpoint_dir) {
+                    Ok(()) => {
+                        info!(target: LOG_CONSENSUS, ?session_checkpoint_dir, ?session_index, "Created db checkpoint");
+                    }
+                    Err(e) => {
+                        warn!(target: LOG_CONSENSUS, ?session_checkpoint_dir, ?session_index, ?e, "Could not create db checkpoint");
+                    }
+                }
+            }
+
+            {
+                // Check if any old checkpoint need to be cleaned up
+                let _timing /* logs on drop */ = timing::TimeReporter::new("remove-database-checkpoint").info();
+                if let Err(e) =
+                    Self::delete_old_database_checkpoints(session_index, &checkpoint_dir)
+                {
+                    warn!(target: LOG_CONSENSUS, ?e, "Could not delete old checkpoints");
+                }
+            }
+        }
+    }
+
+    /// Iterates through the db checkpoint directory and deletes the checkpoint
+    /// equal to `session_index` - `session_diff`.
+    fn delete_old_database_checkpoints(
+        session_index: u64,
+        checkpoint_dir: &Path,
+    ) -> anyhow::Result<()> {
```

Comment:
Thought: should we do the iteration one on start, just to avoid leftovers when changing retention settings?

Reply:
Yes, good idea. Added to `initialize_checkpoint_directory`